configure.ac

```diff
@@ -219,8 +219,13 @@ OMRGLUE=${OMRGLUE:-"./example/glue"}
 ## Default Variables
 
 # Set program defaults if they are undefined
-AC_PROG_CC()
-AC_PROG_CXX()
+# Note: Compilers have already been set up for cross-compiling in the case of RISC-V.
+# Thus, the corresponding detection tests are invalid.
+if test "x$OMR_CROSS_CONFIGURE" != "xyes" -o "x$OMR_HOST_ARCH" != "xriscv";
```

Comment:
> 
> 
> I don't understand why it was necessary to add this test around AC_PROG_CC/CXX. As far as I know, the macros below should work fine with the risc-v cross-compile toolchain. Was there an actual issue that this is solving?

As explained at https://github.com/eclipse/openj9/issues/5058#issuecomment-493135441
This check is useless and misleading in cross-compilation as it has been checked previously in configure and there is no need to do twice.
```
/usr/bin/make -C omr -f run_configure.mk 'SPEC=linux_riscv64_cmprssptrs_cross' 'OMRGLUE=../gc_glue_java' 'CONFIG_INCL_DIR=../gc_glue_java/configure_includes' 'OMRGLUE_INCLUDES=../oti ../include ../gc_base ../gc_include ../gc_stats ../gc_structs ../gc_base ../include ../oti ../nls ../gc_include ../gc_structs ../gc_stats ../gc_modron_standard ../gc_realtime ../gc_trace ../gc_vlhgc' 'EXTRA_CONFIGURE_ARGS='
make[5]: Entering directory '/root/jchau/temp/RISCV_OPENJ9_v2/openj9-openjdk-jdk11/build/linux-riscv64-normal-server-release/vm/omr'
sh configure --disable-auto-build-flag 'OMRGLUE=../gc_glue_java' 'SPEC=linux_riscv64_cmprssptrs_cross'  --enable-OMRPORT_OMRSIG_SUPPORT --enable-OMR_GC --enable-OMR_PORT --enable-OMR_THREAD --enable-OMR_OMRSIG --enable-tracegen --enable-OMR_GC_ARRAYLETS --enable-OMR_GC_DYNAMIC_CLASS_UNLOADING --enable-OMR_GC_MODRON_COMPACTION --enable-OMR_GC_MODRON_CONCURRENT_MARK --enable-OMR_GC_MODRON_SCAVENGER --enable-OMR_GC_CONCURRENT_SWEEP --enable-OMR_GC_SEGREGATED_HEAP --enable-OMR_GC_HYBRID_ARRAYLETS --enable-OMR_GC_LEAF_BITS --enable-OMR_GC_REALTIME --enable-OMR_GC_VLHGC --enable-OMR_PORT_ASYNC_HANDLER --enable-OMR_THR_CUSTOM_SPIN_OPTIONS --enable-OMR_NOTIFY_POLICY_CONTROL --disable-debug 'lib_output_dir=$(top_srcdir)/../lib' 'exe_output_dir=$(top_srcdir)/..' 'GLOBAL_INCLUDES=$(top_srcdir)/../include' --enable-debug --enable-OMR_THR_THREE_TIER_LOCKING --enable-OMR_THR_YIELD_ALG --enable-OMR_THR_SPIN_WAKE_CONTROL --enable-OMRTHREAD_LIB_UNIX --enable-OMR_ARCH_RISCV --enable-OMR_ENV_LITTLE_ENDIAN --enable-OMR_PORT_CAN_RESERVE_SPECIFIC_ADDRESS --enable-OMR_GC_IDLE_HEAP_MANAGER --enable-OMR_GC_TLH_PREFETCH_FTA --enable-OMR_GC_CONCURRENT_SCAVENGER --enable-OMR_GC_ARRAYLETS --host=riscv64-unknown-linux-gnu --enable-OMR_ENV_DATA64 'OMR_TARGET_DATASIZE=64' --enable-OMR_GC_COMPRESSED_POINTERS --enable-OMR_INTERP_COMPRESSED_OBJECT_HEADER --enable-OMR_INTERP_SMALL_MONITOR_SLOT --build=x86_64-pc-linux-gnu 'OMR_CROSS_CONFIGURE=yes' 'AR=riscv64-unknown-linux-gnu-ar' 'AS=riscv64-unknown-linux-gnu-as' 'CC=riscv64-unknown-linux-gnu-gcc --sysroot=/opt/riscv_gnu_toolchain_glibc2.27_v3/sysroot' 'CXX=riscv64-unknown-linux-gnu-g++ --sysroot=/opt/riscv_gnu_toolchain_glibc2.27_v3/sysroot' 'OBJCOPY=riscv64-unknown-linux-gnu-objcopy' libprefix=lib exeext= solibext=.so arlibext=.a objext=.o 'CCLINKEXE=$(CC)' 'CCLINKSHARED=$(CC)' 'CXXLINKEXE=$(CXX)' 'CXXLINKSHARED=$(CXX)' 'OMR_HOST_OS=linux' 'OMR_HOST_ARCH=riscv' 'OMR_TOOLCHAIN=gcc' 'OMR_BUILD_TOOLCHAIN=gcc' 'OMR_TOOLS_CC=gcc' 'OMR_TOOLS_CXX=g++' 'OMR_BUILD_DATASIZE=64' 
checking build system type... x86_64-pc-linux-gnu
checking host system type... riscv64-unknown-linux-gnu
checking OMR_HOST_OS... linux
checking OMR_HOST_ARCH... riscv
checking OMR_TARGET_DATASIZE... 64
checking OMR_TOOLCHAIN... gcc
checking for riscv64-unknown-linux-gnu-gcc... (cached) /opt/riscv_gnu_toolchain_glibc2.27_v3/bin/riscv64-unknown-linux-gnu-gcc
checking whether we are using the GNU C compiler... no  <-------- wrong
checking whether /opt/riscv_gnu_toolchain_glibc2.27_v3/bin/riscv64-unknown-linux-gnu-gcc accepts -g... no
checking for /opt/riscv_gnu_toolchain_glibc2.27_v3/bin/riscv64-unknown-linux-gnu-gcc option to accept ISO C89... unsupported
checking whether we are using the GNU C++ compiler... no  <------ wrong
checking whether riscv64-unknown-linux-gnu-g++ --sysroot=/opt/riscv_gnu_toolchain_glibc2.27_v3/sysroot accepts -g... no
checking for numa.h... no
```
Secondly, there is no easy to modify the code inside AC_PROG_CC/CXX to fix the misleading messages above as the check code there is automatically generated. So, it never hurts to remove the check in this cross-compilation.

Reply:
OK, I would rather find a way to fix the error, rather than disable the tests. can you provide some configure logs from a bad build?

When you say that cc/cxx is being tested twice, is the first test located in openj9/openjdk?

Reply:
> When you say that cc/cxx is being tested twice, is the first test located in openj9/openjdk?

Yes, it shows up when running `configure` before `make`
```
checking resolved symbolic links for CC... no symlink
configure: Using gcc C compiler version 9.2.0 [riscv64-unknown-linux-gnu-gcc (GCC) 9.2.0]
checking for riscv64-unknown-linux-gnu-/opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc... /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc accepts -g... yes
checking for /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc option to accept ISO C89... none needed
checking for riscv64-unknown-linux-gnu-g++... /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-g++
checking resolved symbolic links for CXX... no symlink
configure: Using gcc C++ compiler version 9.2.0 [riscv64-unknown-linux-gnu-g++ (GCC) 9.2.0]
checking whether we are using the GNU C++ compiler... yes

```

Reply:
> OK, I would rather find a way to fix the error, rather than disable the tests. can you provide some configure logs from a bad build?

I am unable to drop the config.log somehow when enabling `AC_PROG_CC/CXX` in OMR.
But the comparison with the code dealing with `AC_PROG_CC/CXX` in jdk11 at  `openj9-openjdk-jdk11/make/autoconf/toolchain.m4` shows it calls `TOOLCHAIN_FIND_COMPILER `before `AC_PROG_CC/CXX`.
```
AC_DEFUN_ONCE([TOOLCHAIN_DETECT_TOOLCHAIN_CORE],
[
  #
  # Setup the compilers (CC and CXX)
  #
  TOOLCHAIN_FIND_COMPILER([CC], [C], $TOOLCHAIN_CC_BINARY) <---
  # Now that we have resolved CC ourself, let autoconf have its go at it
  AC_PROG_CC([$CC])

  TOOLCHAIN_FIND_COMPILER([CXX], [C++], $TOOLCHAIN_CXX_BINARY) <---
  # Now that we have resolved CXX ourself, let autoconf have its go at it
  AC_PROG_CXX([$CXX])
```
where TOOLCHAIN_DETECT_TOOLCHAIN_CORE (implemented in jdk11) is to validate C/C++ compiler in the path.
```
# Try to locate the given C or C++ compiler in the path, or otherwise.
#
# $1 = compiler to test (CC or CXX)
# $2 = human readable name of compiler (C or C++)
# $3 = compiler name to search for
AC_DEFUN([TOOLCHAIN_FIND_COMPILER],
[
...
```
Given that we already picked up the correct cross-compilers in previous configuration, the duplicate check in omr comparably useless for the cross-compilation in our case unless we are forced to implement something similar with TOOLCHAIN_DETECT_TOOLCHAIN_CORE to achieve this.

Reply:
In addition, we need both the cross-compilers (riscv64-unknown-linux-gnu-gcc/g++) and the local compiler gcc/g++ (which are used to locally compile hook/trace gen) work together in the cross-compilation. The check in omr might screw up the combination.

Reply:
In additioin, comparison the config.log in omr with that in JDK11 shows the --sysroot is specified in the check in JDK11 as follows:
e.g.
```
OMR:
configure:4757: checking whether we are using the GNU C compiler
configure:4776: /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc -c   conftest.c >&5
configure:4776: $? = 0
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "OMR"
| #define PACKAGE_TARNAME "omr"
| #define PACKAGE_VERSION "0.1"
| #define PACKAGE_STRING "OMR 0.1"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| #ifndef __GNUC__
|        choke me
| #endif
| 
|   ;
|   return 0;
| }
configure:4785: result: no
```
config.log in jdk11:
```
configure:37589: checking whether we are using the GNU C compiler
configure:37608: /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc -c  --sysroot=/root/jchau/temp/RISCV_OPENJ9_v2/fedora_riscv/fedora_40G/fedora_riscv_mnt  conftest.c >&5
configure:37608: $? = 0
configure:37617: result: yes
```
But I don't think` --sysroot` really matters to affect the result in the check.

Reply:
OK, can you try compiling this test program by hand with the riscv toolchain? What's the compilation error? Does the error go away with the sysroot option?

We shouldn't disable OMR's toolchain detection, because the build system needs to work outside of J9.

Reply:
> OK, can you try compiling this test program by hand with the riscv toolchain? What's the compilation error? Does the error go away with the sysroot option?

There is no difference between with and without `--sysroot`.
```
root@JCHLNXSVL1:~/jchau/temp# /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc  
--sysroot=/root/jchau/temp/RISCV_OPENJ9_v2/fedora_riscv/fedora_40G/fedora_riscv_mnt 
conftest.c
root@JCHLNXSVL1:~/jchau/temp# echo $?
0 <---
root@JCHLNXSVL1:~/jchau/temp# /opt/riscv_gnu_toolchain_master/bin/riscv64-unknown-linux-gnu-gcc 
 conftest.c 
root@JCHLNXSVL1:~/jchau/temp# echo $?
0 <---
```

Reply:
--> We shouldn't disable OMR's toolchain detection, because the build system needs to work outside of J9.

We only disable it in our case (riscv in the cross-compilation).

Reply:
Already talked to @rwy0717,  we just keep the original check here as it never hurts the cross-compilation except the misleading check information in OMR (already verified the cross-build with the check enabled on Fedora/QEMU). It looks like there is problem dealing with cross-compiler in the configure script. Robert will double-check that to see whether there is any solution to the weird issue.

Reply:
An update: I haven't been able to reproduce this issue on OSX, and I don't have a linux VM set up yet. However, I have found some [suspicious behaviour](https://github.com/eclipse/omr/issues/4532) in the configure script.